### DISCUSSION TITLE
Struct vs Dict performance test - Cairo

### DISCUSSION THREAD

**Discussion by danilowhk** [kkrt-labs/kakarot-ssj#79] [General]
# Comparing Updating Struct vs Inserting in Dict: A Benchmark

## Introduction
This report compares the gas cost of maintaining and changing values on a struct against inserting new values into a dictionary (Dict) in Cairo. The tests were conducted using scarb 0.6.0-alpha.2 and cairo 2.1.0-rc2.

Obs: For this test, both initialization of the struct and the dict were added on the tests (Dict initialization cost 15420 gas, and struct initialization didn't change the gas)

## Methodology
The benchmarks were conducted under the following configurations:
- **Language:** Cairo
- **Version:** scarb 0.6.0-alpha.2, cairo 2.1.0-rc2
- **Operations:** 20 different changes/insertions were evaluated
- **Metrics:** Gas consumption

## Overall Result
The comparison highlights a significant difference in gas consumption:
\```
| Approach                | Gas Consumption |
|-------------------------|-----------------|
| Updating Struct         | 57470           |
| Inserting in Dict       | 131040          |
\```

### Description
This approach involves maintaining and changing values on a struct. Each change on the struct is performed 20 times.

### Code
```cairo
#[test]
#[available_gas(1000000)]
fn test_updating_struct(){
    let x = testing::get_available_gas();
    gas::withdraw_gas().unwrap();
    
    
    let mut test_struct = TestStruct { a: 0, b:  2,c:  3, d: 10, e: 20, f:  3, g: 10, h: 20, ab: 1, abc: 2 };
    test_struct.a = test_struct.a + 1;
    test_struct.a = test_struct.a + 1;
    test_struct.a = test_struct.a + 1;
    test_struct.a = test_struct.a + 1;
    test_struct.a = test_struct.a + 1;
    test_struct.a = test_struct.a + 1;
    test_struct.a = test_struct.a + 1;
    test_struct.a = test_struct.a + 1;
    test_struct.a = test_struct.a + 1;        
    test_struct.a = test_struct.a + 1;
    test_struct.a = test_struct.a + 1;
    test_struct.a = test_struct.a + 1;
    test_struct.a = test_struct.a + 1;
    test_struct.a = test_struct.a + 1;
    test_struct.a = test_struct.a + 1;
    test_struct.a = test_struct.a + 1;
    test_struct.a = test_struct.a + 1;
    test_struct.a = test_struct.a + 1;    
    test_struct.a = test_struct.a + 1;
    test_struct.a = test_struct.a + 1;

    '---------Gas Struct---------'.print();
    // 57470 gas for 20 changes, 54820 gas for 19 changes , 2650 gas per change
    (x - testing::get_available_gas()).print();
}
```
### Result
Gas consumption for this test was 57470.

### Description
This approach involves inserting new values into a dictionary (Dict). Each insertion is performed 20 times. An insert on Dict takes around 5000 gas units.

### Code
```cairo
#[test]
#[available_gas(1000000)]
fn test_updating_dict(){
    let x = testing::get_available_gas();
    gas::withdraw_gas().unwrap();    
   let mut balances: Felt252Dict<u64> = Default::default();

    balances.insert(1, balances.get(1) + 1);
    balances.insert(1, balances.get(1) + 1);
    balances.insert(1, balances.get(1) + 1);
    balances.insert(1, balances.get(1) + 1);
    balances.insert(1, balances.get(1) + 1);
    balances.insert(1, balances.get(1) + 1);
    balances.insert(1, balances.get(1) + 1);
    balances.insert(1, balances.get(1) + 1);
    balances.insert(1, balances.get(1) + 1);   
    balances.insert(1, balances.get(1) + 1);
    balances.insert(1, balances.get(1) + 1);
    balances.insert(1, balances.get(1) + 1);
    balances.insert(1, balances.get(1) + 1);
    balances.insert(1, balances.get(1) + 1);
    balances.insert(1, balances.get(1) + 1);
    balances.insert(1, balances.get(1) + 1);
    balances.insert(1, balances.get(1) + 1);
    balances.insert(1, balances.get(1) + 1);   
    balances.insert(1, balances.get(1) + 1);
    balances.insert(1, balances.get(1) + 1);


    '---------Gas Dict---------'.print();
    // 131040 gas for 20 changes, 125430 gas for 19 changes , 5610 gas per change
    (x - testing::get_available_gas()).print();

}
```

### Result
Gas consumption for this test was 131040.

## Conclusion
The tests indicate that maintaining and changing values on a struct (57470 gas) is cheaper than inserting new values into a dictionary (131040 gas) in the given setup and conditions. This result highlights the potential efficiency gains of using structs when handling multiple changes in Cairo.

### Versions used:
scarb: 0.6.0-alpha.2 (4b4d5411b 2023-07-25)
cairo: 2.1.0-rc2 (https://crates.io/crates/cairo-lang-compiler/2.1.0-rc2)

**Comment by Eikix** [comment]
I have a question: are dicts and structs even comparable? 
Why are we comparing them?

**Comment by danilowhk** [comment]
It was to compare bigger Data Types that are "mutable", in our case if we should use one or another to store some data such as program counter , gas , etc..
